From 1e2c984d6b203fb791b2191b62a8b48694e7503c Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Thu, 16 Nov 2017 09:26:55 +0000
Subject: [PATCH 2/2] Add new build option to compile `qol-assist` as a static
 binary

This will allow us to greatly reduce the risk factor to Qol Assist
from upgrades by ensuring it is free of potential library issues from
ongoing upgrades.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 meson.build       | 6 ++++++
 meson_options.txt | 1 +
 src/meson.build   | 7 +++++++
 3 files changed, 14 insertions(+)

diff --git a/meson.build b/meson.build
index 15de99c..9df5e92 100644
--- a/meson.build
+++ b/meson.build
@@ -54,6 +54,11 @@ if with_systemd_system_unit_dir == ''
     endif
 endif
 
+with_static_binary = get_option('with-static-binary')
+
+if with_static_binary == true
+    message('WARNING: Do not build static binaries with glibc!')
+endif
 
 # Sort out config.h now
 cdata = configuration_data()
@@ -104,6 +109,7 @@ report = [
     '    sysconfdir:                             @0@'.format(path_sysconfdir),
     '    sbindir:                                @0@'.format(path_sbindir),
     '    systemd unit directory:                 @0@'.format(with_systemd_system_unit_dir),
+    '    build static binary:                    @0@'.format(with_static_binary),
     '',
     '    Migration tracking:',
     '    ===================',
diff --git a/meson_options.txt b/meson_options.txt
index 5f41307..5cc231f 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -2,3 +2,4 @@ option('with-minimum-uid', type: 'string', description: 'Minimum UID for non-sys
 option('with-wheel-group', type: 'string', description: 'Default wheel (sudo) group name', value: 'sudo')
 option('with-bug-tracker', type: 'string', description: 'Bug tracker URL', value: 'https://dev.solus-project.com')
 option('with-systemd-system-unit-dir', type: 'string', description: 'systemd unit directory')
+option('with-static-binary', type: 'boolean', description: 'Build a static binary for upgrade resilience', value: 'false')
diff --git a/src/meson.build b/src/meson.build
index f7f8ccc..39036ca 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -9,6 +9,12 @@ qol_assist_sources = [
     'util.c',
 ]
 
+# Set static if requested (defaults to being enabled)
+qol_link_args = []
+if with_static_binary == true
+    qol_link_args += '-static'
+endif
+
 # Allow clean definition of migrations
 
 migrations = [
@@ -25,6 +31,7 @@ endforeach
 qol_assist = executable(
     'qol-assist',
     sources: qol_assist_sources,
+    link_args: qol_link_args,
     install: true,
     install_dir: path_sbindir,
     include_directories: [
-- 
2.15.0

